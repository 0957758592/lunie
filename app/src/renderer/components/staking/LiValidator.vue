<template>
  <tr class="data-table__row li-validator">
    <td class="data-table__row__info">
      <img
        v-if="validator.keybase"
        :src="validator.keybase.avatarUrl"
        class="data-table__row__info__image"
        width="48"
        height="48"
      >
      <img
        v-else
        class="
          data-table__row__info__image
          data-table__row__info__image--no-img
        "
        src="~assets/images/validator-icon.svg"
        width="48"
        height="48"
      >
      <div class="data-table__row__info__container">
        <span
          v-tooltip.top="status"
          :class="statusColor"
          class="data-table__row__info__container__status"
        />
        <router-link
          :to="{
            name: 'validator',
            params: { validator: validator.operator_address }
          }"
          :class="styles"
          class="data-table__row__info__container__name"
        >
          {{ validator.description.moniker }}
        </router-link>
        <div class="data-table__row__info__container__description">
          <short-bech32 :address="validator.operator_address" />
        </div>
      </div>
    </td>
    <td class="li-validator__delegated-steak">
      {{ yourVotes.isGreaterThan(0) ? num.shortNumber(yourVotes) : `--` }}
    </td>
    <td class="li-validator__voting-power">
      {{
        validator.percent_of_vote
          ? num.percent(validator.percent_of_vote)
          : `--`
      }}
    </td>
    <td class="li-validator__commission">
      {{ validator.commission ? num.percent(validator.commission) : `--` }}
    </td>
    <td class="li-validator__uptime">
      {{ validator.uptime ? num.percent(validator.uptime) : `--` }}
    </td>
  </tr>
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import { calculateTokens, ratToBigNumber } from "scripts/common"
import ShortBech32 from "common/ShortBech32"
import BigNumber from "bignumber.js"
export default {
  name: `li-validator`,
  components: {
    ShortBech32
  },
  props: {
    validator: {
      type: Object,
      required: true
    },
    disabled: {
      type: Boolean,
      required: true
    }
  },
  data: () => ({ num }),
  computed: {
    ...mapGetters([
      `delegates`,
      `committedDelegations`,
      `distribution`,
      `bondDenom`,
      `session`,
      `lastHeader`
    ]),
    yourVotes() {
      return this.committedDelegations[this.validator.operator_address]
        ? BigNumber(
          num
            .atoms(
              calculateTokens(
                this.validator,
                this.committedDelegations[this.validator.operator_address]
              )
            )
            .toString()
        )
        : BigNumber(0)
    },
    styles() {
      let value = ``
      if (this.validator.isValidator) value += `li-validator-validator `
      return value
    },
    delegateType() {
      return this.validator.revoked
        ? `Revoked`
        : this.validator.isValidator
          ? `Validator`
          : `Candidate`
    },
    powerRatio() {
      return ratToBigNumber(this.validator.tokens)
        .div(this.delegates.globalPower)
        .toNumber()
    },
    status() {
      // status: jailed
      if (this.validator.revoked)
        return `This validator has been jailed and is not currently validating`

      // status: inactive
      if (parseFloat(this.validator.voting_power) === 0)
        return `This validator does not have enough voting power yet and is inactive`

      // status: active
      return `This validator is actively validating`
    },
    statusColor() {
      // status: jailed
      if (this.validator.revoked) return `red`

      // status: inactive
      if (parseFloat(this.validator.voting_power) === 0) return `yellow`

      // status: active
      return `green`
    },
    rewards() {
      if (!this.session.signedIn) return null
      const validatorRewards = this.distribution.rewards[
        this.validator.operator_address
      ]
      return validatorRewards
        ? num.shortNumber(num.atoms(validatorRewards[this.bondDenom]) || 0)
        : null
    }
  },
  watch: {
    lastHeader: {
      immediate: true,
      handler() {
        if (this.yourVotes > 0) {
          this.$store.dispatch(
            `getRewardsFromValidator`,
            this.validator.operator_address
          )
        }
      }
    }
  }
}
</script>
