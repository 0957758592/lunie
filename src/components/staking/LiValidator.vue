<template>
  <tr
    class="li-validator"
    :data-moniker="validator.description.moniker"
    @click="
      $router.push({
        name: 'validator',
        params: { validator: validator.operator_address }
      })
    "
  >
    <td class="data-table__row__info">
      <img
        v-if="validator.keybase && validator.keybase.avatarUrl"
        :src="validator.keybase.avatarUrl"
        class="li-validator-image"
        :alt="`validator logo for ` + validator.description.moniker"
      />
      <img
        v-else
        class="li-validator-image"
        src="~assets/images/validator-icon.svg"
        alt="generic validator logo - graphic triangle supporting atom token"
      />
      <div class="validator-info">
        <h3 class="li-validator-name">{{ validator.description.moniker }}</h3>
        <div v-if="validator.my_delegations > 0">
          <h4>
            {{
              validator.my_delegations
                ? num.shortDecimals(num.atoms(validator.my_delegations))
                : null
            }}
          </h4>
          <h5 v-if="validator.rewards > 0">
            {{
              validator.rewards
                ? `+` + num.shortDecimals(num.atoms(validator.rewards))
                : `--`
            }}
          </h5>
        </div>
      </div>
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'expectedReturns' }">
      {{
        validator.expectedReturns
          ? num.percent(validator.expectedReturns)
          : `--`
      }}
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'voting-power' }">
      {{ validator.tokens ? percentOfVotingPower : `--` }}
    </td>
  </tr>
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import BN from "bignumber.js"
export default {
  name: `li-validator`,
  props: {
    validator: {
      type: Object,
      required: true
    },
    showOnMobile: {
      type: String,
      default: () => "returns"
    }
  },
  data: () => ({ num }),
  computed: {
    ...mapGetters([
      `delegates`,
      `distribution`,
      `session`,
      `lastHeader`,
      `pool`
    ]),
    percentOfVotingPower() {
      return num.percent(
        BN(this.validator.tokens)
          .div(this.pool.pool.bonded_tokens)
          .toFixed(4)
      )
    }
  }
}
</script>
<style scoped>
.li-validator {
  padding: 0.5rem 1rem;
  margin-bottom: 0.25rem;
  border-bottom: 1px solid var(--bc-dim);
  border-radius: 0.25rem;
}

.li-validator:last-child {
  border-bottom: none;
}

.validator-info {
  display: flex;
  flex-direction: column;
  padding-left: 1rem;
  text-overflow: ellipsis;
}

.li-validator h4,
.li-validator h5 {
  font-size: var(--sm);
  display: inline-block;
}

.li-validator h5 {
  color: var(--success);
}

.li-validator:hover {
  cursor: pointer;
  background: var(--hover-bg);
  color: var(--bright);
}

.li-validator-name {
  font-size: 1rem;
  line-height: 18px;
  font-weight: 500;
  white-space: pre-wrap;
  color: var(--bright);
  display: inline-block;
}

.li-validator-image {
  border-radius: 0.25rem;
  height: 2.5rem;
  min-width: 2.5rem;
  border: 1px solid var(--bc-dim);
}
</style>
