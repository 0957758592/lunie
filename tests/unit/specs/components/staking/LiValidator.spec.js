import { shallowMount, createLocalVue } from "@vue/test-utils"
import VueApollo from "vue-apollo"
import LiValidator from "src/components/staking/LiValidator"

const localVue = createLocalVue()
localVue.directive(`tooltip`, () => {})
localVue.use(VueApollo)

describe(`LiValidator`, () => {
  let wrapper, $store

  const validator = {
    //From GraphQL
    avatarUrl: "https://s3.amazonaws.com/img/path.jpg",
    consensus_pubkey: "cosmosvalpub1234",
    customized: false,
    delegator_shares: "697935712090.000000000000000000",
    details: "Mr Mounty",
    id: "9cfd0f08-c750-4cde-a267-e5a38feede20",
    identity: "4BE49EABAA41B8BF",
    jailed: false,
    keybaseId: "4BE49EABAA41B8BF",
    lastUpdated: "2019-08-15T16:03:34.988007+00:00",
    max_change_rate: "0.010000000000000000",
    max_rate: "0.500000000000000000",
    min_self_delegation: "1",
    moniker: "mr_mounty",
    operator_address: "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au",
    profileUrl: "https://keybase.io/melea",
    rate: "0.080000000000000000",
    status: 2,
    tokens: 14,
    unbonding_height: 556130,
    unbonding_time: "2019-06-26T23:50:09.38840326Z",
    update_time: "2019-04-01T17:07:19.892771401Z",
    uptime_percentage: "0.9962",
    userName: "mr_mounty",
    website: "www.monty.ca",

    // Enriched locally
    expectedReturns: 0.13
  }

  beforeEach(() => {
    $store = {
      state: {
        pool: {
          pool: {
            bonded_tokens: 1000
          }
        }
      }
    }

    wrapper = shallowMount(LiValidator, {
      localVue,
      propsData: {
        validator,
        showOnMobile: "returns"
      },
      mocks: {
        $store
      },
      stubs: [`router-link`]
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should show the voting power`, () => {
    expect(wrapper.html()).toContain(`1.40%`)
  })
})
