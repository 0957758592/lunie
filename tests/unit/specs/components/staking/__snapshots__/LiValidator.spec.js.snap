// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiValidator has the expected html structure 1`] = `
<tr
  class="li-validator"
  data-moniker="mr_mounty"
>
  <td
    class="data-table__row__info"
  >
    <apolloquery-stub
      debounce="0"
      options="[object Object]"
      prefetch="true"
      query="[object Object]"
      tag="div"
      throttle="0"
      update="function () { [native code] }"
      variables="[object Object]"
    />
     
    <div
      class="validator-info"
    >
      <h3
        class="li-validator-name"
      >
        
        mr_mounty
      
      </h3>
       
      <!---->
    </div>
  </td>
   
  <td
    class="hide-xs"
  >
    
    13.00%
  
  </td>
   
  <td
    class="hide-xs"
  >
    
    1.40%
  
  </td>
</tr>
`;
